blockchain/My-Wallet-V3-iOS
# feat(HB): Number Entry UX Behavior IOS-1284

## Objective

This is a refactor and implementation of client side logic for exchange amount entry.

## Description

1. `ExchangeInputsAPI` - Lots of changes here. There are now functions for `canDoSomething`, `maxFiatFractional` and `maxAssetFractional`, adding characters or a delimiter (e.g. a comma or decimal point), `backspace`, and returning the correct `NSAttributedString` for either fiat or the given asset. **Note:** You'll see that there's a `ExchangeStyleTemplate`. This has styling data like the correct fonts and colors. This is required for generating the `NSAttributedString` with the correct styling.
2. `InputComponentType` and `InputComponent` - There's a couple of `InputComponentTypes` like `whole` (a whole number), `fractional`, `pendingFractional` (when the user has tapped the decimal point but hasn't entered a fractional amount). `InputComponent` holds the string value and the type.
3. `InputComponent` has an `attributedString`. We stitch an array of `InputComponent` together to make the primary string. 
4. `wigglePrimaryPrimaryLabel` - If the user enters an invalid entry, we trigger haptic feedback and wiggle the label.

**Note: This PR has gone through some merge conflicts so, if you see something that looks like it is old, please call it out.**

## How to Test

1. In `ExchangeCoordinator` call `showExchange(type: .homebrew)` and then `return` in the `start()` function.
2. Build and run
3. Enter in values in the exchange.
4. The values should style correctly
5. Switching the currency type should work.

## Screenshot/Design assessment

N/A

Design Assessment: I'd say this is at 85%.

## Merge Checklist
